docs/develop/cmodules.rst

```diff
@@ -0,0 +1,86 @@
+Extending MicroPython with C
+============================
+
+Some specialized code would be unacceptably slow or needs to access hardware in
+a way that cannot be done from MicroPython. Therefore, it supports a way of
+extending the language with custom modules written in C. But before you consider
+writing a module in C, please take a look at :ref:`speed_python`.
+
+`Unlike CPython <https://docs.python.org/3/extending/building.html>`_, these
+modules are (currently) embedded directly in the program image instead of being
+dynamically loaded. This requires a `custom build of MicroPython
+<https://github.com/micropython/micropython/wiki/Getting-Started>`_.
+
+
+Writing a module
+----------------
+
+A module is a directory with the following files:
+
+  * ``micropython.mk``, which contains the Makefile fragment for this module.
+  * All C files you would like included.
+
+Put the required build commands in ``micropython.mk``. For a simple module, you
+will only have to add the file paths to ``SRC_USERMOD``, which will include
+these C files in the build:
+
+.. highlight:: make
+.. code::
+
+    # Add all C files to SRC_USERMOD.
+    SRC_USERMOD += example/example.c
+
+This is a very bare bones module named ``example`` that provides
+``example.double(x)``. Note that the name of the module must be equal to the
+directory name and is also used in the name of the ``mp_obj_module_t`` object at
+the bottom.
+
+.. highlight:: c
+.. code::
+
+    // Include required definitions first.
+    #include "py/obj.h"
+    #include "py/runtime.h"
+
+    // This is the function you will call using example.double(n).
+    STATIC mp_obj_t example_double(mp_obj_t x_obj) {
+        // Check input value and convert it to a C type.
+        if (!MP_OBJ_IS_SMALL_INT(x_obj)) {
+            mp_raise_ValueError("x is not a small int");
+        }
+        int x = mp_obj_int_get_truncated(x_obj);
```

Comment:
Please don't start with showing users such specialized functions as mp_obj_int_get_truncated(). There's mp_obj_get_int(), which also alleviates from the need to check type.